File: 2015/16.jl
using Combinatorics

struct Sue
    idx::String
    properties::Dict{String, Int}
end

function readInput()
    sues = Vector{Sue}()
    f = open("instructions16a.txt", "r")

    lines = readlines(f)
    for l in lines
        name = split(l, ": ")[1]
        items = split(l, ", ")
        vals = Dict(x[end-1]=>parse(Int, x[end]) for x in split.(items, " "))
        push!(sues, Sue(name, vals))
    end       
    return sues
end

function MFCSAM(sues::Vector{Sue}, outcome)
    remaining = Vector{Sue}()
    for s in sues
        right = true
        for (k, v) in outcome
            if haskey(s.properties, k)
                if s.properties[k] != v
                    right = false
                    break
                end
            end
        end

        if right
            push!(remaining, s)
        end
    end
    println("🎅 $(remaining[1])")
end

function MFCSAM_v2(sues::Vector{Sue}, outcome)
    remaining = Vector{Sue}()
    for s in sues
        right = true
        for (k, v) in outcome
            if haskey(s.properties, k)
                if k == "cats:" || k == "trees:"
                    if s.properties[k] <= v
                        right = false
                        break
                    end
                elseif k == "pomeranians:" || k == "goldfish:"
                    if s.properties[k] >= v
                        right = false
                        break
                    end
                else
                    if s.properties[k] != v
                        right = false
                        break
                    end
                end
            end
        end

        if right
            push!(remaining, s)
        end
    end
    println("🎄 $(remaining[1])")
end

outcome = Dict("children:"=> 3, "cats:"=>7, "samoyeds:"=>2, "pomeranians:"=>3,
               "akitas:"=>0, "vizslas:"=>0, "goldfish:"=>5, "trees:"=>3,
               "cars:"=>2, "perfumes:"=>1)

sues = readInput()
MFCSAM(sues, outcome)
MFCSAM_v2(sues, outcome)

File: 2015/16.py
import time, re

from utils import readInput

r1 = re.compile("(\\d+):")
r2 = re.compile("(\\w+:\\s\\d+)")

class Aunt:
    def __init__(self, name, things):
        self.name = name
        self.things = things

    def check2(self, findings):
        matches = 0
        special = ('cat', 'tree', 'pomeranians', 'goldfish')
        for k, v in findings.items():
            if k in self.things.keys():
                if (k == "cats" or k == "trees") and v < self.things[k]:
                    matches += 1
                elif (k == "pomeranians" or k == "goldfish") and v > self.things[k]:
                    matches += 1
                elif (k not in special) and self.things[k] == v:
                    matches += 1
        return matches

    def check1(self, findings):
        matches = 0
        for k, v in findings.items():
            if k in self.things.keys() and v == self.things[k]:
                matches += 1
        return matches

mfcsam = {'children': 3,
          'cats': 7,
          'samoyeds': 2,
          'pomeranians': 3,
          'akitas': 0,
          'vizslas': 0,
          'goldfish': 5,
          'trees': 3,
          'cars': 2,
          'perfumes': 1}

def loadInput():
    lines = readInput("instructions16a.txt")
    aunts = []
    for d in lines:
        name = r1.findall(d)[0]
        temp = r2.findall(d)
        things = {}
        for k in temp:
            items = k.split(" ")
            things[items[0][:-1]] = int(items[1])
        aunts.append(Aunt(name, things))
    return aunts

def part1(aunts):
    aunt = -1
    best = 0
    for i, a in enumerate(aunts):
        m =  a.check1(mfcsam)
        if m >= best:
            aunt = i
            best = m
    print (f"🎄 Part 1: {aunts[aunt].name} {aunts[aunt].things}")

def part2(aunts):
    aunt = -1
    best = 0
    for i, a in enumerate(aunts):
        m =  a.check2(mfcsam)
        if m >= best:
            aunt = i
            best = m
    print (f"🎄🎅 Part 2: {aunts[aunt].name} {aunts[aunt].things}")
    
if __name__ == "__main__":
    title = "Day 16: Aunt Sue"
    sub = "-"*(len(title)+2)

    print()
    print(f" {title} ")
    print(sub)
    
    inputs = loadInput()
    
    t0 = time.time()
    part1(inputs)
    print ("Time: {:.5f}".format(time.time()-t0))
    
    t0 = time.time()
    part2(inputs)
    print ("Time: {:.5f}".format(time.time()-t0))

